src/histbuf.rs

```diff
@@ -0,0 +1,325 @@
+    pub fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    /// Writes an element to the buffer, overwriting the oldest value.
+    pub fn push(&mut self, t: T) {
+        self.data[self.write_at] = t;
+        self.write_at = (self.write_at + 1) % self.len();
+    }
+
+    /// Clones and pushes all elements in a slice to the buffer.
+    ///
+    /// If the slice is longer than the buffer, only the last `self.len()`
+    /// elements will actually be stored.
+    pub fn extend_from_slice(&mut self, other: &[T])
+    where
+        T: Clone,
+    {
+        for item in other {
+            self.push(item.clone());
+        }
+    }
+
+    /// Returns a reference to the oldest (least recently pushed) value.
+    pub fn first(&self) -> &T {
+        &self.data[self.write_at]
+    }
+
+    /// Returns a reference to the newest (most recently pushed) value.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use heapless::HistoryBuffer;
+    /// use heapless::consts::*;
+    ///
+    /// let mut x: HistoryBuffer<u8, U16> = HistoryBuffer::new();
+    /// x.push(4);
+    /// x.push(10);
+    /// assert_eq!(x.last(), &10);
+    /// ```
+    pub fn last(&self) -> &T {
+        &self.data[(self.write_at + self.len() - 1) % self.len()]
+    }
+
+    /// Returns an iterator over the elements of the buffer.
+    ///
+    /// Note: if the order of elements is not important, use
+    /// `.as_slice().iter()` instead.
+    pub fn iter(&self) -> Iter<'_, T> {
+        Iter {
+            data: &self.data,
+            cur: self.write_at,
+            left: self.len(),
+        }
+    }
+
+    /// Returns an iterator over mutable elements of the buffer.
+    ///
+    /// Note: if the order of elements is not important, use
+    /// `.as_mut_slice().iter_mut()` instead.
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = &'_ mut T> {
+        let (p1, p2) = self.data.split_at_mut(self.write_at);
+        p2.iter_mut().chain(p1)
+    }
+}
+
+pub struct Iter<'a, T> {
+    data: &'a [T],
+    cur: usize,
+    left: usize,
+}
+
+impl<'a, T> Iterator for Iter<'a, T> {
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<&'a T> {
+        if self.left == 0 {
+            None
+        } else {
+            let el = &self.data[self.cur];
+            self.cur = (self.cur + 1) % self.data.len();
+            self.left -= 1;
+            Some(el)
+        }
+    }
+}
+
+impl<T, N> Extend<T> for HistoryBuffer<T, N>
+where
+    N: ArrayLength<T>,
+{
+    fn extend<I>(&mut self, iter: I)
+    where
+        I: IntoIterator<Item = T>,
+    {
+        for item in iter.into_iter() {
+            self.push(item);
+        }
+    }
+}
+
+impl<'a, T, N> Extend<&'a T> for HistoryBuffer<T, N>
+where
+    T: 'a + Clone,
+    N: ArrayLength<T>,
+{
+    fn extend<I>(&mut self, iter: I)
+    where
+        I: IntoIterator<Item = &'a T>,
+    {
+        self.extend(iter.into_iter().cloned())
+    }
+}
+
+impl<T, N> AsSlice for HistoryBuffer<T, N>
```

Comment:
Could you add `as_slice` and `as_mut_slice` as inherent methods so that `hb.as_slice()` works *without* having to import the `AsSlice` trait?

(inherent methods as in `impl HistoryBuffer { fn as_slice(&self) -> &[T] { .. } }`

Reply:
Sure!